lib/mail_room/connection.rb

```diff
@@ -0,0 +1,170 @@
+module MailRoom
+  class Connection
+    def initialize(mailbox)
+      @mailbox = mailbox
+
+      # log in and set the mailbox
+      reset
+      setup
+    end
+
+    def on_new_message(&block)
+      @new_message_handler = block
+    end
+
+    # is the connection logged in?
+    # @return [Boolean]
+    def logged_in?
+      @logged_in
+    end
+
+    # is the connection blocked idling?
+    # @return [Boolean]
+    def idling?
+      @idling
+    end
+
+    # is the imap connection closed?
+    # @return [Boolean]
+    def disconnected?
+      @imap.disconnected?
+    end
+
+    # is the connection ready to idle?
+    # @return [Boolean]
+    def ready_to_idle?
+      logged_in? && !idling?
+    end
+
+    def quit
+      stop_idling
+      reset
+    end
+
+    def wait
+      begin
+        # in case we missed any between idles
+        process_mailbox
```

Comment:
It's the act of uid_fetch that triggers a message to be SEEN for gmail, at least.